qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -383,7 +397,11 @@ def reduce(self, atol: Optional[float] = None, rtol: Optional[float] = None) ->
                 ),
                 spin=self.spin,
             )
-        new_array = flatten_array[non_zero].T.reshape((3, len(non_zero), self.register_length))
+        new_array = (
+            flatten_array[non_zero]
+            .reshape((len(non_zero), 3, self.register_length))
+            .transpose(1, 0, 2)
+        )
```

Comment:
I am not sure whether the bug is fully resolved. For example:
```python
In [1]: from qiskit_nature.operators.second_quantization import SpinOp

In [2]: SpinOp([('+_0 -_0', 1)], register_length=4)
Out[2]: SpinOp([('X_0^2 I_1 I_2 I_3', (1+0j)), ('Y_0^2 I_1 I_2 I_3', (1+0j)), ('Z_0 I_1 I_2 I_3', (1+0j))], spin=1/2, register_length=4)

In [3]: SpinOp([('+_0 -_0', 1)], register_length=4).reduce()
Out[3]: SpinOp([('Z_0 Y_1^2 X_2^2 I_3', (1+0j)), ('I_0 I_1 I_2 I_3', (1+0j)), ('I_0 I_1 I_2 I_3', (1+0j))], spin=1/2, register_length=4)

In [4]: SpinOp([('+_0 -_0', 1)], register_length=4).reduce().reduce()
Out[4]: SpinOp([('I_0 Z_1 I_2 Y_3^2', (2+0j)), ('I_0 X_1^2 I_2 I_3', (1+0j))], spin=1/2, register_length=4)

In [5]: SpinOp([('+_0 -_0', 1)], register_length=4).reduce().reduce().reduce()
Out[5]: SpinOp([('I_0 I_1 Z_2 X_3^2', (2+0j)), ('I_0 I_1 Y_2^2 I_3', (1+0j))], spin=1/2, register_length=4)
```

Calling `reduce()` is still require to be done twice. But at least calling it more often does not accumulate coefficients like it did before.

Reply:
I can not reproduce it. The following is the output of my environment;

<img width="791" alt="image" src="https://user-images.githubusercontent.com/6814928/112559328-194e8b00-8e14-11eb-87fa-445b925fb446.png">

I think you need to update your code since your output is old style. (This PR removed I_{i}.)

Reply:
My apologies, you are absolutely right! I don't quite know what happened here yesterday, but I guess I was a little tired and checkedout the wrong PR.. :see_no_evil: